### PR TITLE
Support for Plutus V2 (ledger API)

### PR DESCRIPTION
**Pull request by JaredCorduan**
These changes add support for the [second version](https://github.com/input-output-hk/plutus/blob/4344fa209aa217c6d16ea7cb70264497a0b9735a/plutus-ledger-api/src/Plutus/V2/Ledger/Api.hs) of the Plutus ledger API (which now includes all the redeemers for a transaction inside the Plutus context). Using `V2` Plutus scripts will result in a phase-one validation error, however, so long as there is no associated cost model.

Subsequent versions (`V3`, etc) should be simpler to add:
  * extend the `Language` enum and follow the GHC warnings.
  * for serialization, add a new key in the witness set (see the [cddl](https://github.com/input-output-hk/cardano-ledger-specs/blob/c6c4be1562e23a3dd48282387c4e48ff918fbab0/eras/shelley/test-suite/cddl-files/shelley.cddl#L200)).

The trace generation and the direct LEDGER property tests are now both using `PlutusV2`.  There is also a unit test to ensure that not having a cost model for `PlutusV2` results in a phase-one error.

### CODE REVIEW

**Comment by michaelpj** [comment]
> Notably missing from this PR is support for per language TxInfo, the information that is passed the Plutus. @michaelpj do you think that all the built in types will change (things like TxOutRef), or just TxInfo?

My plan is to re-export the exact same types for everything except `TxInfo`.

**Comment by michaelpj** [comment]
Okay Jared, you should be able to try this again with plutus master.

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Data.hs [hunk 1]
```diff
@@ -180,19 +177,21 @@ encodeRaw ::
 encodeRaw metadata allScripts =
   Tag 259 $
     Keyed
-      (\m tss pss -> AuxiliaryDataRaw m (StrictSeq.fromList $ tss <> pss))
+      (\m tss p1ss p2ss -> AuxiliaryDataRaw m (StrictSeq.fromList $ tss <> p1ss <> p2ss))
```
Comment (nc6) [inline eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Data.hs]: /immature giggling/

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Data.hs [hunk 2]
```diff
@@ -180,19 +177,21 @@ encodeRaw ::
...
-    getPlutus (PlutusScript x) = Just x
+    getPlutus (PlutusScript _ x) = Just x
     getPlutus _ = Nothing
-    (plutusScripts, timelocks) =
-      List.partition
-        isPlutusScript
-        (Foldable.toList allScripts)
+    sortScripts (ts, v1, v2) s@(TimelockScript _) = (s : ts, v1, v2)
+    sortScripts (ts, v1, v2) s@(PlutusScript PlutusV1 _) = (ts, s : v1, v2)
+    sortScripts (ts, v1, v2) s@(PlutusScript PlutusV2 _) = (ts, v1, s : v2)
```
Comment (nc6) [inline eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Data.hs:192]: Is there a strong reason to serialise plutus scripts in segregated blocks in this way? I just imagine this will get tiresome when we add even more versions.

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Language.hs [hunk 3]
```diff
@@ -33,20 +35,23 @@ instance NFData Language
 
 instance ToCBOR Language where
   toCBOR PlutusV1 = toCBOR (0 :: Int)
```
Comment (nc6) [inline eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Language.hs]: We could use `fromEnum` here to avoid having to update this for every new language, and avoid the possibility of typos.

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Language.hs [hunk 4]
```diff
@@ -33,20 +35,23 @@ instance NFData Language
...
+  toCBOR PlutusV2 = toCBOR (1 :: Int)
 
 instance FromCBOR Language where
   fromCBOR = do
     n <- decodeWord64
     case n of
       0 -> pure PlutusV1
+      1 -> pure PlutusV2
```
Comment (nc6) [inline eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Language.hs]: We could use `toEnum` here for the same reason.

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs [hunk 5]
```diff
@@ -517,6 +517,10 @@ getLanguageView pp lang@PlutusV1 =
   LangDepView
     (serialize' lang)
     (serializeEncoding' $ maybe encodeNull toCBOR $ Map.lookup lang (_costmdls pp))
+getLanguageView pp lang@PlutusV2 =
+  LangDepView
+    (serialize' lang)
+    (serializeEncoding' $ maybe encodeNull toCBOR $ Map.lookup lang (_costmdls pp))
```
Comment (nc6) [inline eras/alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs:523]: Do we expect this implementation to change?

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Tools.hs [hunk 6]
```diff
@@ -47,7 +56,7 @@ data ScriptFailure c
   | -- | Missing datum.
     MissingDatum (DataHash c)
   | -- | Plutus evaluation error.
-    ValidationFailed P.EvaluationError
+    ValidationFailed PV1.EvaluationError [Text]
```
Comment (nc6) [inline eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Tools.hs]: This is rather confusing - what happens to Plutus V2 evaluation errors? I guess the key is that presently these types are the same, but it would be good to clarify this at least in the documentation.

File: libs/plutus-preprocessor/src/Main.hs [hunk 7]
```diff
@@ -12,6 +12,7 @@
 --     cd into the plutus-preprocessor directory and type 'cabal run'
 module Main where
 
+import Cardano.Ledger.Alonzo.Language (Language (..))
```
Comment (nc6) [inline libs/plutus-preprocessor/src/Main.hs:15]: I wonder whether this should live elsewhere. But seems reasonable for now!

**Review by nc6 — DISMISSED** [review]
Few suggested changes, otherwise looks good!

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Tools.hs (hunk 6 above)
Comment (michaelpj) [reply eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Tools.hs]: If we're going to make it very mechanical, then I think we should have different errors here. Or an error GADT indexed by the language enum? :trollface:

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Language.hs (hunk 3 above)
Comment (michaelpj) [reply eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Language.hs]: But then the risk is that reordering the constructors could change serialization, right?

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs [hunk 8]
```diff
@@ -112,12 +113,12 @@ instance NoThunks Tag
 -- | Scripts in the Alonzo Era, Either a Timelock script or a Plutus script.
 data Script era
   = TimelockScript (Timelock (Crypto era))
-  | PlutusScript ShortByteString -- A Plutus.V1.Ledger.Scripts(Script) that has been 'CBOR' encoded
+  | PlutusScript Language ShortByteString
```
Comment (michaelpj) [inline eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs:117]: Hum. I guess I would have expected either:
```
data Script = TimelockScript ... | PlutusV1Script ... | PlutusV2Script ...
```
or
```
data Script = AScript Language ....
```
or maybe even
```
data Script = TimeLockScript ... | NonNativeScript Language ...
```
IDK, there seems like there's maybe still a bit too much "plutus" specificity (if we're trying to be a bit more agnostic about what languages there are).

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs [hunk 9]
```diff
@@ -129,14 +130,14 @@ instance NFData (Script era)
 -- | Both constructors know their original bytes
 instance SafeToHash (Script era) where
   originalBytes (TimelockScript t) = originalBytes t
-  originalBytes (PlutusScript bs) = fromShort bs
+  originalBytes (PlutusScript _ bs) = fromShort bs
 
-alwaysSucceeds, alwaysFails :: Natural -> Script era
-alwaysSucceeds n = PlutusScript (Plutus.alwaysSucceedingNAryFunction n)
-alwaysFails n = PlutusScript (Plutus.alwaysFailingNAryFunction n)
+alwaysSucceeds, alwaysFails :: Language -> Natural -> Script era
```
Comment (michaelpj) [inline eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs]: Hm, I thought we moved these to `cardano-ledger-alonzo-test`...

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs [hunk 10]
```diff
@@ -294,7 +303,8 @@ instance forall era. (Typeable (Crypto era), Typeable era) => ToCBOR (Script era
 
 encodeScript :: (Typeable (Crypto era)) => Script era -> Encode 'Open (Script era)
 encodeScript (TimelockScript i) = Sum TimelockScript 0 !> To i
-encodeScript (PlutusScript s) = Sum PlutusScript 1 !> To s -- Use the ToCBOR instance of ShortByteString
+encodeScript (PlutusScript PlutusV1 s) = Sum (PlutusScript PlutusV1) 1 !> To s -- Use the ToCBOR instance of ShortByteString
```
Comment (michaelpj) [inline eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs:323]: huh, I guess I would have expected this to use the `ToCBOR` instance for Language...

**Review by michaelpj — APPROVED** [review]
I have no idea what I'm talking about, feel free to ignore me.

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Data.hs (hunk 1 above)
Comment (JaredCorduan) [reply eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Data.hs]: oh goodness, I didn't even see that... :facepalm:

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Data.hs (hunk 2 above)
Comment (JaredCorduan) [reply eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Data.hs:192]: I believe this was the original plan all along, but indeed this code is a bit messy, it wasn't pleasant to write...

I'm open to other ideas, but this change does need to be backwards compatible, and we do not currently have the language version in the script serialization.

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Language.hs (hunk 3 above)
Comment (JaredCorduan) [reply eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Language.hs]: That was indeed my worry, but maybe we could just put a comment on the type not to re-order them? I'm happy either way.

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs (hunk 5 above)
Comment (JaredCorduan) [reply eras/alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs:523]: I did not think we could assume that every language view would contain the same data. And I wanted to make sure that GHC would at least make us think about each new language.

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs (hunk 8 above)
Comment (JaredCorduan) [reply eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs:117]: I was going to cross that bridge when we got there, I don't really know how to anticipate what shape other languages will have.

Comment (JaredCorduan) [reply eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs:117]: That said, I'm open to suggestions if you have a favorite!

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs (hunk 9 above)
Comment (JaredCorduan) [reply eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs]: good point, these don't belong here.

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Tools.hs (hunk 6 above)
Comment (JaredCorduan) [reply eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Tools.hs]: good catch!

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs (hunk 10 above)
Comment (JaredCorduan) [reply eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs:323]: I know, it's a bit awkward. See https://github.com/input-output-hk/cardano-ledger-specs/pull/2485/files#r717306756

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs [hunk 11]
```diff
@@ -129,27 +131,31 @@ instance NFData (Script era)
...
 data ExUnits = ExUnits
-  { exUnitsMem :: !Word64,
-    exUnitsSteps :: !Word64
+  { exUnitsMem :: !Natural,
+    exUnitsSteps :: !Natural
   }
   deriving (Eq, Generic, Show)
   -- It is deliberate that there is no Ord instance, use `pointWiseExUnits` instead.
   deriving
-    (BoundedMeasure, Measure)
```
Comment (kk-hainq) [inline eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs:152]: @JaredCorduan May I ask why was this removed? `AlonzoMeasure` in `ouroboros-network` wants it:
https://github.com/input-output-hk/ouroboros-network/blob/e0ccbb73296027a5e5fd224a78374321974758ce/ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Mempool.hs#L314

Comment (JaredCorduan) [reply eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs:152]: We were waffling about how best to make this instance, so to move this PR along I removed the instance. This PR adds it back in, with one proposed solution: https://github.com/input-output-hk/cardano-ledger-specs/pull/2515

Comment (kk-hainq) [reply eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs:152]: Awesome, thank you for the reference!
